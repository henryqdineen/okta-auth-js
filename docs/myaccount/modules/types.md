[@okta/okta-auth-js/myaccount](../README.md) / [Modules](../modules.md) / types

# Module: types

## Table of contents

### Enumerations

- [EmailRole](../enums/types.EmailRole.md)
- [Status](../enums/types.Status.md)

### Classes

- [BaseTransaction](../classes/types.BaseTransaction.md)
- [EmailChallengeTransaction](../classes/types.EmailChallengeTransaction.md)
- [EmailStatusTransaction](../classes/types.EmailStatusTransaction.md)
- [EmailTransaction](../classes/types.EmailTransaction.md)
- [PhoneTransaction](../classes/types.PhoneTransaction.md)
- [ProfileSchemaTransaction](../classes/types.ProfileSchemaTransaction.md)
- [ProfileTransaction](../classes/types.ProfileTransaction.md)

### Type aliases

- [AddEmailPayload](types.md#addemailpayload)
- [AddPhonePayload](types.md#addphonepayload)
- [ChallengePhonePayload](types.md#challengephonepayload)
- [EmailProfile](types.md#emailprofile)
- [IAPIFunction](types.md#iapifunction)
- [MyAccountRequestOptions](types.md#myaccountrequestoptions)
- [PhoneProfile](types.md#phoneprofile)
- [UpdateProfilePayload](types.md#updateprofilepayload)
- [VerificationPayload](types.md#verificationpayload)

## Type aliases

### AddEmailPayload

Ƭ **AddEmailPayload**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `profile` | { `email`: `string`  } |
| `profile.email` | `string` |
| `role` | [`EmailRole`](../enums/types.EmailRole.md) |
| `sendEmail` | `boolean` |

#### Defined in

[types.ts:32](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L32)

___

### AddPhonePayload

Ƭ **AddPhonePayload**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `method` | `string` |
| `profile` | { `phoneNumber`: `string`  } |
| `profile.phoneNumber` | `string` |
| `sendCode` | `boolean` |

#### Defined in

[types.ts:46](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L46)

___

### ChallengePhonePayload

Ƭ **ChallengePhonePayload**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `method` | `string` |

#### Defined in

[types.ts:54](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L54)

___

### EmailProfile

Ƭ **EmailProfile**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `email` | `string` |

#### Defined in

[types.ts:28](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L28)

___

### IAPIFunction

Ƭ **IAPIFunction**<`T`\>: (`oktaAuth`: `OktaAuthInterface`, `options?`: [`MyAccountRequestOptions`](types.md#myaccountrequestoptions)) => `Promise`<`T`\>

#### Type parameters

| Name |
| :------ |
| `T` |

#### Type declaration

▸ (`oktaAuth`, `options?`): `Promise`<`T`\>

##### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

##### Returns

`Promise`<`T`\>

#### Defined in

[types.ts:13](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L13)

___

### MyAccountRequestOptions

Ƭ **MyAccountRequestOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `accessToken?` | `string` |
| `challengeId?` | `string` |
| `emailId?` | `string` |
| `id?` | `string` |
| `payload?` | [`AddEmailPayload`](types.md#addemailpayload) \| [`AddPhonePayload`](types.md#addphonepayload) \| [`ChallengePhonePayload`](types.md#challengephonepayload) \| [`VerificationPayload`](types.md#verificationpayload) \| [`UpdateProfilePayload`](types.md#updateprofilepayload) |

#### Defined in

[types.ts:72](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L72)

___

### PhoneProfile

Ƭ **PhoneProfile**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `profile` | { `phoneNumber`: `string`  } |
| `profile.phoneNumber` | `string` |

#### Defined in

[types.ts:40](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L40)

___

### UpdateProfilePayload

Ƭ **UpdateProfilePayload**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `profile` | { `[property: string]`: `any`; `email?`: `string` ; `firstName?`: `string` ; `lastName?`: `string` ; `login?`: `string`  } |
| `profile.email?` | `string` |
| `profile.firstName?` | `string` |
| `profile.lastName?` | `string` |
| `profile.login?` | `string` |

#### Defined in

[types.ts:62](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L62)

___

### VerificationPayload

Ƭ **VerificationPayload**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `verificationCode` | `string` |

#### Defined in

[types.ts:58](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/types.ts#L58)
