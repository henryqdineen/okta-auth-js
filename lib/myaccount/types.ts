import { OktaAuthInterface } from '../types';

export { 
  EmailTransaction, 
  EmailStatusTransaction,
  EmailChallengeTransaction,
  PhoneTransaction,
  ProfileTransaction,
  ProfileSchemaTransaction,
  BaseTransaction
} from './transactions';

export type IAPIFunction<T> = (
  oktaAuth: OktaAuthInterface, 
  options?: MyAccountRequestOptions
) => Promise<T>;

export enum EmailRole {
  PRIMARY = 'PRIMARY',
  SECONDARY = 'SECONDARY'
}

export enum Status {
  VERIFIED = 'VERIFIED',
  UNVERIFIED = 'UNVERIFIED'
}

export type EmailProfile = {
  email: string;
}

export type AddEmailPayload = {
  profile: {
    email: string;
  };
  sendEmail: boolean;
  role: EmailRole;
}

export type PhoneProfile = {
  profile: {
    phoneNumber: string;
  };
}

export type AddPhonePayload = {
  profile: {
    phoneNumber: string;
  };
  sendCode: boolean;
  method: string;
};

export type ChallengePhonePayload = {
  method: string;
}

export type VerificationPayload = {
  verificationCode: string;
};

export type UpdateProfilePayload = {
  profile: {
    firstName?: string;
    lastName?: string;
    email?: string;
    login?: string;
    [property: string]: any;
  };
};

export type MyAccountRequestOptions = {
  id?: string;
  emailId?: string;
  challengeId?: string;
  payload?: AddEmailPayload 
    | AddPhonePayload 
    | ChallengePhonePayload
    | VerificationPayload 
    | UpdateProfilePayload;
  accessToken?: string;
}
