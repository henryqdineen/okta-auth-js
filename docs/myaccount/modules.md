[@okta/okta-auth-js/myaccount](README.md) / Modules

# @okta/okta-auth-js/myaccount

## Table of contents

### Modules

- [index](modules/index.md)
- [types](modules/types.md)
