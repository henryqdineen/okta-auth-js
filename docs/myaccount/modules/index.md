[@okta/okta-auth-js/myaccount](../README.md) / [Modules](../modules.md) / index

# Module: index

## Table of contents

### Functions

- [addEmail](index.md#addemail)
- [addPhone](index.md#addphone)
- [deleteEmail](index.md#deleteemail)
- [deletePhone](index.md#deletephone)
- [getEmail](index.md#getemail)
- [getEmailChallenge](index.md#getemailchallenge)
- [getEmails](index.md#getemails)
- [getPhone](index.md#getphone)
- [getPhones](index.md#getphones)
- [getProfile](index.md#getprofile)
- [getProfileSchema](index.md#getprofileschema)
- [sendEmailChallenge](index.md#sendemailchallenge)
- [sendPhoneChallenge](index.md#sendphonechallenge)
- [updateProfile](index.md#updateprofile)
- [verifyEmailChallenge](index.md#verifyemailchallenge)
- [verifyPhoneChallenge](index.md#verifyphonechallenge)

## Functions

### addEmail

▸ **addEmail**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

#### Defined in

[emailApi.ts:45](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L45)

___

### addPhone

▸ **addPhone**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

#### Defined in

[phoneApi.ts:44](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L44)

___

### deleteEmail

▸ **deleteEmail**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[emailApi.ts:63](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L63)

___

### deletePhone

▸ **deletePhone**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:62](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L62)

___

### getEmail

▸ **getEmail**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

#### Defined in

[emailApi.ts:28](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L28)

___

### getEmailChallenge

▸ **getEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

#### Defined in

[emailApi.ts:96](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L96)

___

### getEmails

▸ **getEmails**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)[]\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)[]\>

#### Defined in

[emailApi.ts:12](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L12)

___

### getPhone

▸ **getPhone**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

**`scope:`** okta.myaccount.phone.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

#### Defined in

[phoneApi.ts:27](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L27)

___

### getPhones

▸ **getPhones**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)[]\>

**`scope:`** okta.myaccount.phone.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)[]\>

#### Defined in

[phoneApi.ts:11](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L11)

___

### getProfile

▸ **getProfile**(`oktaAuth`, `options?`): `Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

**`scope:`** okta.myaccount.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

#### Defined in

[profileApi.ts:11](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L11)

___

### getProfileSchema

▸ **getProfileSchema**(`oktaAuth`, `options?`): `Promise`<[`ProfileSchemaTransaction`](../classes/types.ProfileSchemaTransaction.md)\>

**`scope:`** okta.myaccount.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileSchemaTransaction`](../classes/types.ProfileSchemaTransaction.md)\>

#### Defined in

[profileApi.ts:42](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L42)

___

### sendEmailChallenge

▸ **sendEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

#### Defined in

[emailApi.ts:79](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L79)

___

### sendPhoneChallenge

▸ **sendPhoneChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:78](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L78)

___

### updateProfile

▸ **updateProfile**(`oktaAuth`, `options?`): `Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

**`scope:`** okta.myaccount.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

#### Defined in

[profileApi.ts:24](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L24)

___

### verifyEmailChallenge

▸ **verifyEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[emailApi.ts:113](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L113)

___

### verifyPhoneChallenge

▸ **verifyPhoneChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:95](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L95)
