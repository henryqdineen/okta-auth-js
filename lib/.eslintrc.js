module.exports = {
  extends: ["plugin:compat/recommended"],
  env: {
    browser: false,
    node: false
  },
  settings: {
    polyfills: [
      "Promise",
      "Array.from",
      "TextEncoder",
      "Object.assign"
    ]
  },
  parserOptions: {
    sourceType: "module",
    ecmaVersion: 2020
  },
  rules: {
    "node/no-unsupported-features/es-syntax": 0,
    "@typescript-eslint/no-non-null-assertion": 0,
    "no-use-before-define": 0,
    "@typescript-eslint/no-unused-vars": [
      "error", 
      { 
        argsIgnorePattern: "^_"
      }
    ],
    // duplicates with @typescript-eslint/no-unused-vars
    "no-unused-vars": "off"
  }
}
