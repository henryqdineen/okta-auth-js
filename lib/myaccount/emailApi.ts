import { sendRequest } from './request';
import { 
  IAPIFunction, 
  BaseTransaction, 
  EmailTransaction, 
  EmailChallengeTransaction 
} from './types';

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmails: IAPIFunction<EmailTransaction[]> = async (
  oktaAuth,
  options?
) => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction[];
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'GET',
    accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const addEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
): Promise<EmailTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'POST',
    payload,
    accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const deleteEmail: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'DELETE',
    accessToken
  }) as BaseTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const sendEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}/challenge`,
    method: 'POST',
    accessToken,
    transactionClassName: 'EmailChallengeTransaction'
  }) as EmailChallengeTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { emailId, challengeId, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}`,
    method: 'POST',
    accessToken,
    transactionClassName: 'EmailChallengeTransaction'
  }) as EmailChallengeTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const verifyEmailChallenge: IAPIFunction<BaseTransaction> = async (
  oktaAuth,
  options
) => {
  const { emailId, challengeId, payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}/verify`,
    method: 'POST',
    payload,
    accessToken
  }) as BaseTransaction;
  return transaction;
};
