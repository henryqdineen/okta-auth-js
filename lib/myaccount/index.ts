export * from './profileApi';
export * from './emailApi';
export * from './phoneApi';
