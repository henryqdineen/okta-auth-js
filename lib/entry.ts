import OktaAuth from './OktaAuth';
import * as MyAccountMethods from './myaccount';
import { OktaAuthOptions } from './types';

class OktaAuthCDN extends OktaAuth {
  myaccount;

  constructor(args: OktaAuthOptions) {
    super(args);

    this.myaccount = Object.entries(MyAccountMethods)
      .filter(([ name ]) => name !== 'default')
      .reduce((acc, [name, fn]) => {
        acc[name] = (fn as any).bind(null, this);
        return acc;
      }, {});
  }
}

export default OktaAuthCDN;
