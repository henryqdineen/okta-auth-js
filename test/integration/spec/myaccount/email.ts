import {
  addEmail,
  deleteEmail,
  getEmail,
  getEmails,
  sendEmailChallenge,
  verifyEmailChallenge
} from '../../../../lib/myaccount';
import {
  createClient,
  signinAndGetTokens
} from '../../util';
import { EmailRole } from '../../../../lib/myaccount/types';
import { AuthApiError } from '../../../../lib';
import { EmailTransaction } from '../../../../lib/myaccount/transactions';

// TODO: generate dynamic profile with a18n api
describe('MyAccount Email API', () => {
  let client, token;

  beforeAll(async () => {
    client = createClient({});
    const {
      tokens: {
        accessToken: { accessToken } = {}
      }
    } = await signinAndGetTokens(client, {
      scopes: [
        'openid',
        'profile',
        'okta.myAccount.email.read',
        'okta.myAccount.email.manage'
      ]
    });
    token = accessToken;
  });

  describe('getEmail/s', () => {
    it('can get email/s', async () => {
      let email;
      const transactions = await getEmails(client, {
        accessToken: token
      });
      for (const transaction of transactions) {
        expect(transaction).toMatchSnapshot({
          headers: expect.any(Object),
          _http: {
            headers: expect.any(Object),
          },
          id: expect.any(String),
          profile: {
            email: expect.any(String)
          },
          roles: expect.any(Array),
          status: expect.any(String)
        });

        // assign emailTransaction for later test use
        email = transaction;
      }

      // get email by id
      const transaction = await getEmail(client, {
        accessToken: token,
        id: email.id
      });
      expect(transaction).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        },
        roles: expect.any(Array),
        status: expect.any(String)
      });
    });
  });

  describe('Email Management', () => {
    let email: EmailTransaction;

    afterEach(async () => {
      if (email) {
        try {
          await email.delete();
        } catch {
          // do nothing
        }
      }
    });

    it('can manage email with Transaction functions', async () => {
      // create test email
      email = await addEmail(client, {
        accessToken: token,
        payload: {
          profile: {
            email: 'fake1@acme.com'
          },
          sendEmail: false,
          role: EmailRole.SECONDARY
        }
      });


      expect(email).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        }
      });

      // get email
      const getTransaction = await email.get();
      expect(getTransaction).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        }
      });

      const challenge = await email.challenge();
      expect(challenge).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        expiresAt: expect.any(String),
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        }
      });

      const emailStatus = await challenge.poll();
      expect(emailStatus).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        expiresAt: expect.any(String),
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        }
      });

      try {
        // use invalid code to test the api. expect error
        // TODO: integrate with a18n api to test against real verification code
        await challenge.verify({ verificationCode: '000000' });
      } catch (err) {
        expect((err as AuthApiError).xhr?.status).toEqual(401);
      }

      // delete created email
      const deleteTransaction = await email.delete();
      expect(deleteTransaction._http.status).toEqual(204);
    });

    it('can manage email with SDK methods', async () => {
      // create test email
      email = await addEmail(client, {
        accessToken: token,
        payload: {
          profile: {
            email: 'fake2@acme.com'
          },
          sendEmail: false,
          role: EmailRole.SECONDARY
        }
      });
      expect(email).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        }
      });

      // send challenge code
      const challenge = await sendEmailChallenge(client, {
        id: email.id,
        accessToken: token,
      });
      expect(challenge).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        expiresAt: expect.any(String),
        id: expect.any(String),
        profile: {
          email: expect.any(String)
        }
      });

      // verify challenge
      try {
        // use invalid code to test the api. expect error
        // TODO: integrate with a18n api to test against real verification code
        await verifyEmailChallenge(client, {
          accessToken: token,
          emailId: email.id,
          challengeId: challenge.id,
          payload: { verificationCode: '000000' }
        });
      } catch (err) {
        expect((err as AuthApiError).xhr?.status).toEqual(401);
      }

      // delete created email
      const transaction = await deleteEmail(client, {
        id: email.id,
        accessToken: token
      });
      expect(transaction._http.status).toEqual(204);
    });

  });

});

