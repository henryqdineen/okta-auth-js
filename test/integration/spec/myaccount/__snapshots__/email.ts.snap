// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccount Email API Email Management can manage email with SDK methods 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 201,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "email": Any<String>,
  },
  "roles": Array [
    "SECONDARY",
  ],
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Email API Email Management can manage email with SDK methods 2`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 201,
  },
  "expiresAt": Any<String>,
  "headers": Any<Object>,
  "id": Any<String>,
  "poll": [Function],
  "profile": Object {
    "email": Any<String>,
  },
  "status": "UNVERIFIED",
  "verify": [Function],
}
`;

exports[`MyAccount Email API Email Management can manage email with Transaction functions 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 201,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "email": Any<String>,
  },
  "roles": Array [
    "SECONDARY",
  ],
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Email API Email Management can manage email with Transaction functions 2`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "email": Any<String>,
  },
  "roles": Array [
    "SECONDARY",
  ],
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Email API Email Management can manage email with Transaction functions 3`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 201,
  },
  "expiresAt": Any<String>,
  "headers": Any<Object>,
  "id": Any<String>,
  "poll": [Function],
  "profile": Object {
    "email": Any<String>,
  },
  "status": "UNVERIFIED",
  "verify": [Function],
}
`;

exports[`MyAccount Email API Email Management can manage email with Transaction functions 4`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "expiresAt": Any<String>,
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "email": Any<String>,
  },
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Email API getEmail/s can get email/s 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "email": Any<String>,
  },
  "roles": Any<Array>,
  "status": Any<String>,
}
`;

exports[`MyAccount Email API getEmail/s can get email/s 2`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "email": Any<String>,
  },
  "roles": Any<Array>,
  "status": Any<String>,
}
`;
